### PR TITLE
Concurrency headers on create/update/delete

### PR DESCRIPTION
**Pull request by leplatrem**
* [x] Safe creations
* [x] Safe deletions
* [x] Safe updates
* [x] Rework tests

### CODE REVIEW

File: src/reducers/bucket.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ import { omit } from "../utils";
 const INITIAL_STATE: BucketState = {
   busy: false,
   id: null,
+  last_modified: null,
```
Comment (n1k0) [inline src/reducers/bucket.js]: How about we keep it in the `data` object above instead?

File: src/reducers/bucket.js [hunk 2]
```diff
@@ -40,11 +41,12 @@ function load(state: BucketState, bucket: BucketResource): BucketState {
...
   }
   const {data, permissions} = bucket;
-  const {id} = data;
+  const {id, last_modified} = data;
   return {
     ...state,
     busy: false,
     id,
+    last_modified,
     data: omit(data, ["id", "last_modified"]),
```
Comment (n1k0) [inline src/reducers/bucket.js]: Maybe we should not omit `last_modified` here?

File: src/components/bucket/BucketEdit.js [hunk 3]
```diff
@@ -7,17 +7,20 @@ import BucketTabs from "./BucketTabs";
 
 export default class BucketEdit extends Component {
   deleteBucket = (bid) => {
-    const {deleteBucket} = this.props;
+    const {deleteBucket, bucket} = this.props;
+    const {last_modified} = bucket;
     if (confirm("This will delete the bucket and all the collections and " +
                 "records it contains. Are you sure?")) {
-      deleteBucket(bid);
+      deleteBucket(bid, last_modified);
```
Comment (n1k0) [inline src/components/bucket/BucketEdit.js]: I'm uncomfortable doing this in a view component; I suspect we could inspect the resource data state from the store in a saga to infer what should be sent to the server?

File: src/sagas/bucket.js [hunk 4]
```diff
@@ -50,11 +50,11 @@ export function* updateBucket(getState, action) {
 }
 
 export function* deleteBucket(getState, action) {
-  const { bid } = action;
+  const { bid, last_modified } = action;
```
Comment (n1k0) [inline src/sagas/bucket.js]: Note that from here, current bucket last_modified could be easily retrieved from `getState().bucket.data`.

Comment (n1k0) [inline src/sagas/bucket.js]: Scratch it, we're speaking deletion here, so we need the last_modified being passed in the action. This comment still actually applies for updates though.

**Comment by leplatrem** [comment]
We could generalize the approach for delete operations too. The only exception would be the deletion from the list of records, which would need to pass the last_modified via the action. Should I go on with delete?

**Comment by n1k0** [comment]
> The only exception would be the deletion from the list of records, which would need to pass the last_modified via the action

Yes, delete actions need to accept an optional last_modified

File: src/sagas/bucket.js [hunk 5]
```diff
@@ -37,10 +37,13 @@ export function* createBucket(getState, action) {
 
 export function* updateBucket(getState, action) {
   const {bid, bucketData} = action;
+  const {bucket: currentBucket} = getState();
+  const {last_modified} = currentBucket;
+  const savedBucket = {...bucketData, last_modified};
```
Comment (n1k0) [inline src/sagas/bucket.js]: Nit: this gives the false impression the bucket is already saved; `updatedBucket`?

**Pull request merged**